File: 108-将有序数组转换为二叉搜索树.cpp
class Solution {
public:
    TreeNode* sorted_array_to_bst(vector<int>& nums, int left, int right){
        if(left > right)
            return NULL;

        if(left == right){
            TreeNode* root = new TreeNode(nums[left]);
            return root;
        }

        int middle = left + (right - left) / 2;
        TreeNode* root = new TreeNode(nums[middle]);

        root->left = sorted_array_to_bst(nums, left, middle - 1);
        root->right = sorted_array_to_bst(nums, middle + 1, right);

        return root;
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sorted_array_to_bst(nums, 0, nums.size() - 1);
    }
};

File: 108-将有序数组转换为二叉搜索树.py
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:

        def sorted_array_to_bst(nums, left, right):
            if left > right:
                return None

            if left == right:
                return TreeNode(nums[left])
            
            middle = (left + right) // 2
            root = TreeNode(nums[middle])

            root.left = sorted_array_to_bst(nums, left, middle - 1)
            root.right = sorted_array_to_bst(nums, middle + 1, right)

            return root
        
        return sorted_array_to_bst(nums, 0, len(nums) - 1)
